[fix] a2a tasks

RayST3 opened the pull request:
## Summary

Adding get task and cancel tasks

## Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [x] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

---

## Additional Notes

Add any important context (deployment instructions, screenshots, security considerations, etc.)

VirusDumb committed:
Tasks

VirusDumb committed:
add cancel tasks

VirusDumb committed:
format and validate

VirusDumb committed:
Merge branch 'main' of https://github.com/agno-agi/agno into fix-a2a-tasks

VirusDumb committed:
validation fix

VirusDumb committed:
Update router.py

VirusDumb committed:
Update router.py

VirusDumb committed:
Update router.py

VirusDumb committed:
Update router.py

VirusDumb committed:
Update router.py

manuhortet committed:
Merge branch 'main' into fix-a2a-tasks

manuhortet commented on libs/agno/agno/os/interfaces/a2a/router.py:
```diff
@@ -39,6 +41,7 @@ def attach_routes(
     agents: Optional[List[Agent]] = None,
     teams: Optional[List[Team]] = None,
     workflows: Optional[List[Workflow]] = None,
+    dbs: Optional[dict[str, list[Union[BaseDb, AsyncBaseDb]]]] = None,
```
Not Optional, should never be None

manuhortet commented on libs/agno/agno/os/interfaces/a2a/router.py:
```diff
@@ -239,6 +246,124 @@ async def a2a_stream_agent(request: Request, id: str):
+        request_body = await request.json()
+        request_id = request_body.get("id", "unknown")
+        task_id = request_body.get("params", {}).get("task_id")
+        session_id = request_body.get("params", {}).get("session_id")
```
This is the endpoint to get the list of tasks for the given agent (`/agents/{id}/v1/tasks`), so:

- We don't need `task_id` in the request body
- Let's rename the function to `get_agent_tasks`

manuhortet commented on libs/agno/agno/os/interfaces/a2a/router.py:
```diff
@@ -239,6 +246,124 @@ async def a2a_stream_agent(request: Request, id: str):
+        if not session:
+            raise HTTPException(status_code=404, detail=f"Session with ID {session_id} not found")
+
+        runs = session.get("runs")  # type: ignore
```
Let's check why `# type ignore`. It's okay if this is None

manuhortet commented on libs/agno/agno/os/interfaces/a2a/router.py:
```diff
@@ -239,6 +246,124 @@ async def a2a_stream_agent(request: Request, id: str):
+        "/agents/{id}/v1/tasks",
+        operation_id="get_agent_task",
+        name="get_agent_task",
+        description="Retrieve an A2A Task by its ID.",
```
```suggestion
        description="Retrieve the list of A2A Tasks for the contextual Agent",
```

manuhortet reviewed (commented):
The comments for the Agent endpoints apply also to the Team and Workflow endpoints

VirusDumb committed:
Update libs/agno/agno/os/interfaces/a2a/router.py

Co-authored-by: manu <manuhortet@gmail.com>

VirusDumb committed:
Update libs/agno/agno/os/interfaces/a2a/router.py

Co-authored-by: manu <manuhortet@gmail.com>

VirusDumb committed:
Update router.py

VirusDumb committed:
Update reasoning_agent.py

VirusDumb committed:
Update router.py

VirusDumb committed:
Update router.py

manuhortet committed:
Merge branch 'main' into fix-a2a-tasks

manuhortet commented on libs/agno/agno/os/interfaces/a2a/router.py:
```diff
@@ -239,6 +246,244 @@ async def a2a_stream_agent(request: Request, id: str):
+            "jsonrpc": "2.0",
+            "id": request_id,
+            "result": a2a_tasks,
+        }
```
We are first collecting all runs, then iterating over the full list to map them to a2a tasks. 
We can simplify and just track the tasks:

```python
all_tasks = []

...

all_tasks.extend([map_run_schema_to_a2a_task(run) for run in runs]
```

manuhortet commented on libs/agno/agno/os/interfaces/a2a/router.py:
```diff
@@ -239,6 +246,244 @@ async def a2a_stream_agent(request: Request, id: str):
+        id: str,
+        session_type: Optional[SessionType] = None,
+        user_id: Optional[str] = None,
+    ):
```
Let's add the type signature: `-> dict`

manuhortet commented on libs/agno/agno/os/interfaces/a2a/utils.py:
```diff
@@ -283,6 +284,211 @@ def map_run_output_to_a2a_task(run_output: Union[RunOutput, WorkflowRunOutput])
     )
 
 
+def map_run_schema_to_a2a_task(run_schema: Union[RunSchema, TeamRunSchema, WorkflowRunSchema]) -> Task:
```
run_schema will always be a dictionary here, i think

manuhortet commented on libs/agno/agno/os/interfaces/a2a/utils.py:
```diff
@@ -283,6 +284,211 @@ def map_run_output_to_a2a_task(run_output: Union[RunOutput, WorkflowRunOutput])
+    if messages:
+        for msg in messages:
+            if isinstance(msg, dict):
+                role = Role.agent if msg.get("role") == "assistant" else Role.user
```
We need to filter out messages with `role`s that are not `assistant` or `user`

manuhortet commented on libs/agno/agno/os/interfaces/a2a/utils.py:
```diff
@@ -283,6 +284,211 @@ def map_run_output_to_a2a_task(run_output: Union[RunOutput, WorkflowRunOutput])
+
+                message_history.append(
+                    A2AMessage(
+                        message_id=msg.get("id") or str(uuid4()),
```
the message should always have an id. We should not create one for messages that alreaday exist

manuhortet committed:
Merge branch 'main' into fix-a2a-tasks

VirusDumb commented on libs/agno/agno/os/interfaces/a2a/router.py:
```diff
@@ -239,6 +246,244 @@ async def a2a_stream_agent(request: Request, id: str):
+    ):
+        """List all tasks for an Agent. If session_id is provided in params, returns tasks from that session only."""
+        request_body = await request.json()
+        request_id = request_body.get("id", "unknown")
```
should i remove the default?

VirusDumb committed:
Refactor A2A task mapping and artifact creation

Updated the router to return mapped A2A tasks and improved the mapping logic in utils.py. Artifact IDs now use UUIDs, and artifact names and descriptions are derived from URLs and metadata. Message role assignment and ID handling are more explicit, and default values are removed for run_id and session_id.

VirusDumb committed:
Merge branch 'main' into fix-a2a-tasks

VirusDumb committed:
Merge branch 'main' of https://github.com/agno-agi/agno into fix-a2a-tasks

VirusDumb committed:
Merge branch 'fix-a2a-tasks' of https://github.com/agno-agi/agno into fix-a2a-tasks

VirusDumb committed:
Update router.py

VirusDumb committed:
Update router.py

VirusDumb committed:
Merge branch 'main' into fix-a2a-tasks

VirusDumb committed:
Update router.py

VirusDumb committed:
Merge branch 'fix-a2a-tasks' of https://github.com/agno-agi/agno into fix-a2a-tasks

VirusDumb committed:
update utils.py

VirusDumb committed:
Update router.py

VirusDumb committed:
Update router.py

VirusDumb committed:
Update test_a2a.py

VirusDumb committed:
Update test_a2a.py

VirusDumb committed:
Merge branch 'main' into fix-a2a-tasks

manuhortet committed:
Merge branch 'main' into fix-a2a-tasks

VirusDumb committed:
Update utils.py

VirusDumb committed:
Merge branch 'fix-a2a-tasks' of https://github.com/agno-agi/agno into fix-a2a-tasks

VirusDumb committed:
Update router.py

VirusDumb committed:
Update utils.py

VirusDumb committed:
Update utils.py

VirusDumb committed:
Update utils.py

manuhortet committed:
feat: add response schemas for Task endpoints

manuhortet committed:
feat: update cancel endpoints

manuhortet committed:
feat: type list endpoints results

manuhortet committed:
feat: clean-up utils

manuhortet committed:
feat: add tests

manuhortet committed:
Merge branch 'main' into fix-a2a-tasks

VirusDumb committed:
fix cancellation logic and add pagination

VirusDumb committed:
Merge branch 'main' into fix-a2a-tasks

uzaxirr committed:
Merge branch 'main' into fix-a2a-tasks

manuhortet committed:
Merge branch 'main' into fix-a2a-tasks

manuhortet committed:
update

manuhortet committed:
Merge branch 'main' into fix-a2a-tasks

VirusDumb committed:
Update router.py

VirusDumb committed:
Merge branch 'main' into fix-a2a-tasks

VirusDumb committed:
Merge branch 'main' into fix-a2a-tasks

ysolanky committed:
Merge branch 'main' into fix-a2a-tasks
